### PR TITLE
executor: introduce setWithMemoryUsage to track set memory in AggExec.

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:
Many partial result of distinct aggregate contains a Set, For example 
```
type partialResult4AvgDistinctDecimal struct {
	partialResult4AvgDecimal
	valSet set.StringSet
}
```
The memory usage can't be tracked.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Introduce setWithMemoryUsage in package aggfuncs, to track the memory usage of set.

I will replace all Set to setWithMemoryUsage in the next PR.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- No release note

### CODE REVIEW

**Comment by wshwsh12** [comment]
Benchmark Result

Benchmark tracks all memory allocation, but doesn't minus GC memory. So in this cases, Map memory use will be 2 times.
Float64
BenchmarkFloat64SetMemoryUsage/MapRows_0-16         	18133598	        76.9 ns/op	      48 B/op	       1 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_100-16       	   76440	     14942 ns/op	    3199 B/op	      19 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_10000-16     	     795	   1570211 ns/op	  389611 B/op	     257 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_1000000-16   	       6	 170320706 ns/op	49817870 B/op	   38266 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_851968-16    	      12	 103182141 ns/op	27388881 B/op	   38336 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_851969-16    	       9	 114823948 ns/op	49667440 B/op	   38329 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_425984-16    	      18	  60039823 ns/op	13597523 B/op	   19086 allocs/op
BenchmarkFloat64SetMemoryUsage/MapRows_425985-16    	      18	  59216265 ns/op	24739645 B/op	   19098 allocs/op
When rownum is 1000000, B will be 18, and the estimated memory is (1<<18)*(8*(1+8+0)+16) = 23068672. In benchmark result, the result is 49817870.

Int64
BenchmarkInt64SetMemoryUsage/MapRows_0-16         	16186348	        77.6 ns/op	      48 B/op	       1 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_100-16       	  109587	     11015 ns/op	    3198 B/op	      19 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_10000-16     	    1105	   1082723 ns/op	  389633 B/op	     257 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_1000000-16   	       8	 125177175 ns/op	49833701 B/op	   38387 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_851968-16    	      12	  86841121 ns/op	27388939 B/op	   38356 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_851969-16    	      12	  93875048 ns/op	49674116 B/op	   38355 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_425984-16    	      27	  45411134 ns/op	13601213 B/op	   19089 allocs/op
BenchmarkInt64SetMemoryUsage/MapRows_425985-16    	      25	  40704813 ns/op	24740725 B/op	   19093 allocs/op
When rownum is 1000000, B will be 18, and the estimated memory is (1<<18)*(8*(1+8+0)+16) = 23068672. In benchmark result, the result is 49833701.

String
BenchmarkStringSetMemoryUsage/MapRows_0-16         	17873625	        79.9 ns/op	      48 B/op	       1 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_100-16       	   90524	     13844 ns/op	    5424 B/op	      10 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_10000-16     	     769	   1915759 ns/op	  715097 B/op	   10115 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_1000000-16   	       4	 278824461 ns/op	93452812 B/op	 1038068 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_851968-16    	       6	 177443102 ns/op	52171016 B/op	  890108 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_851969-16    	       6	 169181730 ns/op	92271781 B/op	  890059 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_425984-16    	      13	  86480077 ns/op	25957139 B/op	  445012 allocs/op
BenchmarkStringSetMemoryUsage/MapRows_425985-16    	      12	  85994027 ns/op	45993578 B/op	  444891 allocs/op
When rownum is 1000000, B will be 18, and the estimated memory is (1<<18)*(8*(1+16+0)+16) = 39845888. In benchmark result, the result is 93452812.

**Review by ichn-hu — APPROVED** [review]

File: executor/aggfuncs/set_with_memory_usage.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
```
Comment (lzmhhh123) [inline executor/aggfuncs/set_with_memory_usage.go:14]: It's better to move the package to `util/set`

Comment (wshwsh12) [reply executor/aggfuncs/set_with_memory_usage.go]: Addressed. PTAL again

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**